### PR TITLE
Feat/rslad distillation compression

### LINKED ISSUE #47: [Feature]: Implement RSLAD-style robust distillation in compression phase
## Problem Statement

The compression phase (`CompressionPhase` in `nightmarenet/training/phases.py`) applies magnitude pruning but does NOT perform the "robust distillation" step described in the README:

> **Compress** - Preserve robustness, shed parameters. Adversarial robust distillation (RSLAD-style) + magnitude pruning.

The actual implementation only does pruning + optional fine-tuning:

```python
class CompressionPhase:
    def run(self, dataloader, optimizer):
        # 1. Apply magnitude pruning
        pruner = MagnitudePruner(pruning_ratio=self.pruning_ratio)
        stats = pruner.apply(self.model)
        # 2. Optional fine-tune after prune
        if self.finetune_after_prune:
            # ... standard fine-tuning loop ...
```

There is no knowledge distillation from the pre-pruned model to the pruned model. The README's claim of "RSLAD-style adversarial robust distillation" is unimplemented. This is significant because:

- Standard pruning followed by fine-tuning does NOT preserve adversarial robustness (known finding from Ye et al. 2019)
- The distillation step is what makes the compression phase part of the robustness mechanism (not just a size reduction)
- The paper draft references this but the code doesn't implement it

**Relevant files:**
- `nightmarenet/training/phases.py` (CompressionPhase.run)
- `nightmarenet/compression/pruning.py` (MagnitudePruner only)

## Proposed Solution

Add a distillation step to CompressionPhase:

1. Before pruning, snapshot the current model as the "teacher"
2. After pruning, run a distillation loop where the pruned "student" learns from the teacher's logits
3. Use KL divergence loss between teacher and student outputs (following RSLAD: use adversarial examples as distillation data, not clean data)
4. Config: `compression.distillation: true`, `compression.distillation_temperature: 4.0`, `compression.distillation_alpha: 0.7`

**Files affected:**
- `nightmarenet/compression/pruning.py` or new `nightmarenet/compression/distillation.py`
- `nightmarenet/training/phases.py` (CompressionPhase)
- `configs/default.yaml` (new config keys)

**Dependencies:** None new (KL divergence is in torch.nn.functional).

## Acceptance Criteria

- [ ] CompressionPhase performs KL distillation from pre-pruned teacher to pruned student
- [ ] Distillation uses adversarially-perturbed inputs (not clean data) per RSLAD
- [ ] Config keys added: `compression.distillation`, `compression.distillation_temperature`, `compression.distillation_alpha`
- [ ] Distillation is optional (backward compatible when disabled)
- [ ] Test: verify distillation reduces KL divergence between teacher and student outputs
- [ ] Documentation: update README compress phase description if behavior changes

## Scope / Difficulty Estimate

Medium-high. ~4-5 hours. Requires understanding the distillation literature and implementing a training loop within the compression phase.

### PR DESCRIPTION
**Pull request by RitikaSen27**
## Summary

Implements RSLAD-style adversarial robust distillation in the CompressionPhase to recover robustness lost during magnitude pruning.

## Motivation

Magnitude pruning zeroes out weights without any mechanism to recover the robustness that is lost in the process. This PR addresses issue #47 by snapshotting the pre-pruned model as a frozen teacher and training the pruned student via KL divergence loss on FGSM-perturbed inputs (RSLAD-style), recovering robustness after compression.

Closes #47 

## Changes

- nightmarenet/compression/distillation.py — new module implementing fgsm_perturb (adversarial input generation using torch.autograd.grad) and run_distillation (RSLAD-style KL distillation loop)
- nightmarenet/training/phases.py — teacher snapshot taken before pruning; distillation runs after pruning; distillation_stats included in return dict
- configs/default.yaml — added opt-in config keys: distillation (false by default), distillation_temperature, distillation_alpha, distillation_epochs

-

## Acceptance Criteria

- [x] Teacher is snapshotted from the pre-pruned model and frozen
- [x] Distillation runs on FGSM-adversarially perturbed inputs
- [x] KL divergence loss used between teacher and student logits with temperature scaling
- [x] Feature is opt-in (distillation: false by default) and backward compatible
- [x] All 492 existing tests pass

## Type

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] Feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would break existing behavior)
- [ ] Refactor (no functional change)
- [ ] Documentation
- [ ] Tests
- [ ] Infrastructure (CI/CD, Docker, tooling)

---

## Pre-submission Requirements

> These are mandatory. PRs missing any item will not be reviewed.

- [x] I have starred this repository
- [x] I have followed @Adit-Jain-srm
- [x] I have read CONTRIBUTING.md in full

## Quality Checklist

- [x] ruff check nightmarenet/ — 0 errors
- [x] mypy nightmarenet/ — no new errors
- [x] pytest tests/ — all 492 tests pass
- [x] New functionality has corresponding tests
- [x] Commit messages follow Conventional Commits
- [x] No from __future__ import annotations added under nightmarenet/api/
- [x] No new imports of hosted-only libraries in nightmarenet/`

## Documentation

- [ ] No documentation needed (test-only or internal refactor)
- [ ] README.md updated
- [ ] `docs/` updated (architecture, research, or API docs)
- [x] Docstrings added/updated (Google style, public APIs only)
- [x] Config comments updated (configs/default.yaml)
- [ ] CHANGELOG entry added (if user-facing change)

## AI Disclosure

- [ ] This PR is entirely human-written
- [x] This PR includes AI-assisted code (Claude) — I have reviewed every line and can explain it

## Security Considerations

- [x] Not applicable (no security-sensitive changes)
- [ ] User input is validated before use
- [ ] No secrets, keys, or credentials in code or comments
- [ ] CORS / auth / rate limiting behavior unchanged (or explicitly documented)
- [ ] If this is a security fix, see [SECURITY.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/SECURITY.md) — do NOT describe the vulnerability publicly until patched

## Breaking Changes

N/A

---

## Summary by CodeRabbit

* **New Features**
  * Added optional teacher-student distillation during model compression.
  * Introduced adversarially perturbed training to improve compressed model robustness.
  * Added configurable distillation temperature, weighting, and training epochs.
  * Compression results now include distillation training statistics.

* **Configuration**
  * Simplified pruning settings and added controls for enabling and tuning distillation.

### CODE REVIEW

**Comment by RitikaSen27** [comment]
Hi @Adit-Jain-srm, I have addressed all the CodeRabbit review comments and updated the PR:

Fixed fgsm_perturb to use torch.autograd.grad instead of loss.backward() to prevent student parameter gradients from being corrupted during adversarial example generation

Changed distillation default to false in configs/default.yaml so the feature is truly opt-in and backward compatible

Included distillation_stats in the CompressionPhase.run return dict so metrics are surfaced correctly

Fixed all ruff lint errors (import order, from __future__ import annotations, trailing newline)

All 492 tests pass locally with ruff check nightmarenet/ showing zero errors and mypy nightmarenet/ clean

AI disclosure: This PR includes AI-assisted code generation. I have reviewed every line and can explain it.
Requesting re-review. Thank you!

**Review by Adit-Jain-srm — CHANGES_REQUESTED** [review]
You opened a new PR instead of committing to previous one, not sure if intended. Please close either one on them and commit the latest code to the remaining one after code-rabbit's latest comments resolution; I'll merge it then.
Great work! @RitikaSen27

**Comment by RitikaSen27** [comment]
Closed PR #174 as PR #198  has all the latest fixes including CodeRabbit review resolutions. Please review #198. @Adit-Jain-srm

**Review by Adit-Jain-srm — APPROVED** [review]
LGTM!
